### ISSUE TITLE
Ember 4.0 Tracking Issue

### ISSUE THREAD

**Issue by MelSumner** [emberjs/ember.js#19545]
# Summary
The purpose of this issue is to provide a way to track the status of the work necessary to ship v.4.0 of Ember.js 

The list of issues should be updated to include links to individual issues as they are created. The individual issues should contain more details for each item on this list. 

As a reminder: Major versions of Ember do not ship new features- instead, major versions remove previously deprecated features. To read more about Ember's philosophy regarding _editions_, please see https://emberjs.com/editions/. 

# Issues

- [x] **blocking** Remove deprecated features (See #19617 for open tasks) (includes IE11 removals)
- [x] [Tracking for Remove jQuery by Default](https://github.com/emberjs/rfc-tracking/issues/3) - Done from the Ember 4.0 perspective.
- [x] :lock: `Ember.assign` deprecation implementation https://github.com/emberjs/ember.js/pull/19649
- [x] **blocking** Deprecation guide for `Ember.assign`
- [x] :lock: Ember `auto` location deprecation implementation https://github.com/emberjs/ember.js/pull/19510 - we got this in targeting 4.1
- [x] _non-blocking for 4.0_ :lock: Deprecation guide for `auto` https://github.com/ember-learn/deprecation-app/pull/882
- [ ] _non-blocking:_ [Remove classic examples from API docs](https://github.com/emberjs/ember.js/issues/19702)
- [ ] **blocking** [Make sure we aren't showing deprecated methods in the guides](https://github.com/ember-learn/guides-source/issues/1711) - mostly done except willTransition/didTransition
- [x] **blocking** [Update the optional features guide](https://github.com/ember-learn/guides-source/issues/1712) - has a PR
- [x] **blocking** [Update the Octane upgrade guide](https://github.com/ember-learn/guides-source/issues/1713) - has a PR
- [x] Land https://github.com/emberjs/ember.js/pull/19825 otherwise newly generated files will error due to the newer version of `eslint-plugin-qunit` in the 4.0 new project blueprint 
- [x] [New browser-support policy](https://github.com/emberjs/rfcs/blob/master/text/0685-new-browser-support-policy.md): One to drop IE11 from Ember Data CI.
- [x] **blocking** [New browser-support policy](https://github.com/emberjs/rfcs/blob/master/text/0685-new-browser-support-policy.md): Need a canonical place to publish supported browsers, and we need to have the list for 4.0 in hand. And write it into the blog post.
- [x] **blocking** The implicit injection of the store, which is deprecated, was not deprecated in all cases. @igorT has seen an OSS codebase that broke on 4.0 upgrade because the case was missed. This should be addressed as a release bugfix improving the fidelity of the deprecation. https://github.com/emberjs/data/pull/7744

## Communication

Some of this may happen after release, and is implicitly _non-blocking_.

- [x] Ember.js website updates. Audit the situation and approve by checking this box.
- [x] Add 4.x section to the deprecation guide https://github.com/ember-learn/deprecation-app/issues/898
- [x] Make sure API docs work. They should after release, confirm by checking this box.
- [x] Make sure the Super Rentals app is happy. Check this box the day after we release the final beta of 4.0 and have a green build. https://github.com/ember-learn/super-rentals-tutorial/actions/runs/1451232432
- [ ] Ember 4.0 release blog post [tracking issue](https://github.com/ember-learn/ember-blog/issues/1013) and [Pull request](https://github.com/ember-learn/ember-blog/pull/1036)

**Comment by simonihmig** [comment]
@MelSumner you might want to link to https://github.com/emberjs/rfc-tracking/issues/3 for the jQuery removal, which already lists the work to be done for Ember 4.0! Some PRs already underway... 🙂

**Comment by MelSumner** [comment]
@simonihmig added, thank you!

**Comment by simonihmig** [comment]
Here's a PR for dropping IE11: https://github.com/emberjs/ember.js/pull/19558 :tada:

**Comment by nlfurniss** [comment]
@MelSumner [this PR](https://github.com/ember-learn/deprecation-app/pull/899) can be linked to `Deprecation guide for Ember.assign` 👍

**Comment by jenweber** [comment]
Hi all, I made some new issues for tracking docs work and the blog post. I added them to the main tracking list above. The learning team is filling in details for these issues over the coming weeks.

**Comment by jenweber** [comment]
Another status update, I have a PR open for adding placeholders for missing deprecation guides. These are missing from the 3.x series and not the new deprecations for v4. However we really want these links to be in place in the source code for 4.0.x. https://github.com/ember-learn/deprecation-app/pull/954

**Comment by jenweber** [comment]
I added a new point to the tracking list: the deprecation workflow isn't catching everything, therefore our story of "resolve all deprecations and you can upgrade smoothly" doesn't work. [Bug] Some Deprecations seem to side-step the deprecation workflow #19753

**Comment by mansona** [comment]
I found another thing that could possibly go in the Bugs/Blockers section: https://github.com/emberjs/ember-render-modifiers/pull/33

Essentially we [recommend the use of ember-render-modifiers in the guides here](https://guides.emberjs.com/release/upgrading/current-edition/glimmer-components/#toc_dom-manipulation) but that PR is blocking it from being used in Ember Beta (4.0). I don't have permission to edit the tracking issue so could someone else add it for me?

**Comment by jrjohnson** [comment]
Pulling this comment over from the data PR - now that store is not injected the behavior [documented](https://guides.emberjs.com/release/routing/specifying-a-routes-model/#toc_dynamic-models) for dynamic models as:
> If you do not define a model hook for a route, it will default to using Ember Data to look up the record, as shown below:

no longer works because there is no store. 

It confusingly [errors with the message](https://github.com/jrjohnson/reproduction-data-no-injection) 
> Assertion Failed: _RouteName_ has no method "find"

I think this behavior of automatically using the ED store to lookup the model when no model hook is defined needs to be deprecated or else the store needs to be looked up from the container in the default `model()` hook on routes where it is needed. Otherwise apps with no deprecation warnings which use the current implicit model behavior will error.

**Comment by mixonic** [comment]
@jrjohnson that error around 'has no method find" should be improved in Ember 4.0.1. Some links:

* https://github.com/emberjs/ember.js/pull/19858
* https://github.com/emberjs/ember.js/pull/19860

If you explicitly inject a store in 4.0 and have no model hook, implicit record loading from data will still occur. I'm not sure if it is deprecated yet but several people would like to. Regardless, that itself is not a 4.0 release issue. It is merely a poor API we want to deprecate for 5.0.

I hope that closes the loop, I know your comment here was already a bit stale but didn't want to leave it hanging.

**Comment by mixonic** [comment]
FWIW we suggest (as in, the assert messages suggests) adding a no-op model hook over an injection. 👍

Agreed it is weird, but the messaging should be acceptable. In 4.0.1+ with no route, you get the improved assert I linked above telling you to add a model hook. If you use the ember generate commands, you get a model hook by default if there is a dynamic segment.

This isn't perfect, I agree we should deprecate the implicit loading entirely, but that needs to aim at 5.0.

Additionally, there should be a deprecation in ember 3.28.7+ when the implicitly injected store was accessed. This was an improvement on the earlier deprecation added specifically to address this exact situation. See https://github.com/emberjs/ember.js/pull/19854 I expect your app would have hit that deprecation.

Perhaps you upgraded before that point release was out? It is definitely something we identified from people upgrading in the wild.

However, there is no breaking change besides the removals. Yes, if you created a new route in ember/data 3.28 with a dynamic segment and no route, the behavior is different than if you do the same in 4.0. That is not a breaking change for existing code (that should trigger the above linked deprecation), it is simply a change of behavior for new code.
